apollo-router/src/plugins/traffic_shaping/mod.rs

```diff
@@ -375,16 +379,23 @@ impl TrafficShaping {
                 tower::retry::RetryLayer::new(retry_policy)
             });
 
+            let timeout = config.shaping.timeout.unwrap_or(DEFAULT_TIMEOUT);
             Either::A(ServiceBuilder::new()
-
                 .option_layer(config.shaping.deduplicate_query.unwrap_or_default().then(
                   QueryDeduplicationLayer::default
                 ))
-                    .layer(TimeoutLayer::new(
-                        config.shaping
-                        .timeout
-                        .unwrap_or(DEFAULT_TIMEOUT),
-                    ))
+                    .map_future_with_request_data(
+                        |req: &subgraph::Request| req.context.clone(),
+                        move |ctx, response| {
+                            request_timeout(timeout, response).unwrap_or_else(|error| {
```

Comment:
I think naming is confusing here. It's not really a `response`  it's a future and `request_timeout` maybe we should rename it for something like `check_timeout` or `execute_with_timeout`. I don't know. But it would make the review easier

Reply:
Fixed ✅